FWCore/Services/plugins/AsyncService.cc

```diff
@@ -0,0 +1,47 @@
+#include "FWCore/Concurrency/interface/Async.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
+
+#include <atomic>
+
+namespace edm::service {
+  class AsyncService : public Async {
+  public:
+    AsyncService(ParameterSet const& iConfig, ActivityRegistry& iRegistry);
+
+    static void fillDescriptions(ConfigurationDescriptions& descriptions);
+
+  private:
+    void ensureAllowed() const final;
+
+    std::atomic<bool> allowed_ = true;
+  };
+
+  AsyncService::AsyncService(ParameterSet const& iConfig, ActivityRegistry& iRegistry) {
+    iRegistry.watchPreSourceEarlyTermination([this](TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreGlobalEarlyTermination([this](GlobalContext const&, TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreStreamEarlyTermination([this](StreamContext const&, TerminationOrigin) { allowed_ = false; });
```

Comment:
Should we also add `postEndJob`?

Reply:
I see it would be consistent. I could do it already in `preEndJob`, but then the exception message should be improved, and in a way that would be step towards restricting more when the `AsyncService` can be used that I decided to leave for a later PR. I'll add the `postEndJob`.